Implement `actionable_blocked_uri?` checks

jacobbednarz opened the pull request:
Lays the groundwork for filtering out unactionable violation reports. The idea
here is that not every violation report you receive is actually going to be
something you have control over so you just drop it on the ground and not clog
up your reporting tooling with it.

This implements the first lot of checks based on the `blocked-uri` key of the
report. This will check against a list of known unactionable URIs and halt the
job should one be encountered. The list is compiled from all over the web but
notable mentions to report-uri.io and @oreoshake's [blog article about
Twitter's CSP collector design](https://oreoshake.github.io/csp/twitter/2014/07/25/twitters-csp-report-collector-design.html).

jacobbednarz committed:
Move JSON into job

JSON is only used in the resque job so there is no need to be requiring
it in the app.rb

jacobbednarz committed:
Add known unactionable "block-uri"

This is a list of the blocked URIs that are outside of your control and
is best to ignore otherwise you waste your days trying to fix them.

jacobbednarz committed:
Implement `actionable_blocked_uri?` checks

Add a start point for filtering out bogus requests generated by known
useless 'blocked-uri' directives. This will ensure that if a report
comes in and contains one of the dodgey directives, we just drop it on
the floor instead of trying to process it.

jacobbednarz committed:
s/KNOWN_UNACTIONABLE_REPORT_URIS/KNOWN_UNACTIONABLE_BLOCKED_URIS

oreoshake commented on jobs/create_csp_violation_job.rb:
```diff
@@ -1,7 +1,45 @@
+  # are generated by things like browser extensions, internal browser pages,
+  # MITM proxies and are outside of our control.
+  def self.actionable_blocked_uri?
+    !KNOWN_UNACTIONABLE_BLOCKED_URIS.include?(@violation_report['blocked-uri'])
```
You'd need to parse out the scheme or treat `KNOWN_UNACTIONABLE_BLOCKED_URIS` as a set of regexen, right? IIRC there were things after e.g. `safari-resource://kdjerhrgkljhdekjvhskjehfkuhe`

jacobbednarz replied:
Yep - Good catch! I had this around the wrong way in my head. I've just pushed up a change to use `start_with?` and the known unactionable URI's via 3d898d4. This should now work as intended.

jacobbednarz committed:
Fix logic for `actionable_blocked_uri?`

In 51c25fc the logic that took the `KNOWN_UNACTIONABLE_BLOCKED_URIS`
and then checked if it included the incoming blocked-uri was backwards
and resulted in incorrect results. I've swapped out that logic for a
call to `start_with?` and a list of the known unactionable URIs and it
will now work as expected by comparing the beginning of the blocked URI
with the list of known unactionable URI's we've defined.

oreoshake commented on jobs/create_csp_violation_job.rb:
```diff
@@ -1,7 +1,43 @@
+  # are generated by things like browser extensions, internal browser pages,
+  # MITM proxies and are outside of our control.
+  def self.actionable_blocked_uri?
+    !@violation_report['blocked-uri'].start_with?(
```
:cool: TIL `start_with?` accepts varargs.

jacobbednarz replied:
yep! i think this is nicer than a giant, messy regex.
